intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c

```diff
@@ -8395,47 +8422,125 @@ static void ne_compute_forward_flash_attn_f32_f16_f16(const struct ne_compute_pa
   const int64_t seq_past = seq_all - seq_cur;
   const int64_t batch = neq3;
 
-  const int step_k_bs = k->nb[3] / sizeof(float);
-  const int step_v_bs = v->nb[3] / sizeof(float);
-
   if (params->type == NE_TASK_INIT) {
     return;
   }
 
   if (params->type == NE_TASK_FINALIZE) {
     return;
   }
+  const int keles = ne_element_size(k);
+  const int veles = ne_element_size(v);
+  int step_k_sl = k->nb[1] / keles;
+  int step_k_head_num = k->nb[2] / keles;
+  int step_k_head_size = k->nb[0] / keles;
+  int step_k_bs = k->nb[3] / keles;
+  int step_v_sl = v->nb[0] / veles;
+  int step_v_head_size = v->nb[1] / veles;
+  int step_v_head_num = v->nb[2] / veles;
+  int step_v_bs = k->nb[3] / veles;
   float scale = *(float*)dst->padding;
   bool mask = *(bool*)&dst->padding[sizeof(scale)];
   attn_fp32_fp16_fp16_fp32_fwd_args_t args = {
       .Q = (float*)q->data,
       .K = (ne_fp16_t*)k->data,
       .V = (ne_fp16_t*)v->data,
-      .batch_size = batch,
-      .head_size = headsize,
-      .head_num = headnum,
       .dst = (float*)dst->data,
-      .is_causal = mask,
+      .Q_sc = 1.f,
+      .K_sc = 1.f,
+      .V_sc = 1.f,
+      .dst_sc = 1.f,
+      .tmp = tmp->data,
       .QK_scale = scale,
+      .is_causal = mask,
+      .batch_size = batch,
+      .head_num = headnum,
+      .head_size = headsize,
       .sl_q = seq_cur,
       .sl_kv = seq_all,
+      .Q_layout = ATTN_FWD_LAYOUT_PLAIN,
+      .K_layout = ATTN_FWD_LAYOUT_PLAIN,
+      .V_layout = ATTN_FWD_LAYOUT_PLAIN,
+      .dst_layout = ATTN_FWD_LAYOUT_PLAIN,
       .step_q_bs = seq_cur * embedsize,
       .step_q_head_num = headsize,
       .step_q_sl = embedsize,
       .step_k_bs = step_k_bs,
-      .step_k_head_num = headsize,
-      .step_k_sl = embedsize,
+      .step_k_head_num = step_k_head_num,
+      .step_k_sl = step_k_sl,
+      .step_k_head_size = step_k_head_size,  // TODO
       .step_v_bs = step_v_bs,
-      .step_v_head_num = headsize,
-      .step_v_sl = embedsize,
+      .step_v_head_num = step_v_head_num,
+      .step_v_sl = step_v_sl,
+      .step_v_head_size = 1,
       .step_dst_bs = seq_cur * embedsize,
       .step_dst_head_num = headsize,
       .step_dst_sl = embedsize,
-      .tmp = tmp->data,
   };
   jblas_fusion_attn_fp32_fp16_fp16_fp32_forward(&args);
 }
 
+static void ne_compute_forward_flash_attn_reordered(const struct ne_compute_params* params, const struct ne_tensor* q,
```

Comment:
can you move all functions you updated to another file? seems ne_layers.c is much larger than expected

Reply:
Agree that the `ne_layers.c` is too large.

If we move the implementation to a new file (say "ne_layers_mha.c"), will we also move their declarations to a new header file, or keep then in `ne_layers.h`?

Reply:
headers might not change

Reply:
`ne_xxx` and `ne_compute_xxx` functions are put in `ne_layers.c`, while detailed functions for actual checking & computing are declared in `layers/mha_dense.h` and implemented in `layers/mha_dense.cpp`